compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -316,51 +324,63 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
 
         for (nll_error, _) in nll_errors.into_iter() {
             match nll_error {
-                RegionErrorKind::TypeTestError { type_test } => {
+                // A type-test failed and the constraint was rewritten due
+                // to higher-ranked trait bounds.
+                RegionErrorKind::TypeTestError {
+                    generic_kind,
+                    lower_bound,
+                    span,
+                    failed_due_to_placeholders,
+                } if failed_due_to_placeholders => {
+                    // FIXME. We should handle this case better. It
+                    // indicates that we have e.g., some region variable
+                    // whose value is like `'a+'b` where `'a` and `'b` are
+                    // distinct unrelated universal regions that are not
+                    // known to outlive one another. It'd be nice to have
+                    // some examples where this arises to decide how best
+                    // to report it; we could probably handle it by
+                    // iterating over the universal regions and reporting
+                    // an error that multiple bounds are required.
+                    let mut diag = self.dcx().create_err(GenericDoesNotLiveLongEnough {
+                        kind: generic_kind.to_string(),
+                        span,
+                    });
+
+                    // Add notes and suggestions for the case of 'static lifetime
+                    // implied but not specified when a generic associated types
+                    // are from higher-ranked trait bounds
+                    self.suggest_static_lifetime_for_gat_from_hrtb(&mut diag, lower_bound);
+
+                    self.buffer_error(diag);
+                }
+                RegionErrorKind::TypeTestError {
+                    lower_bound,
+                    span,
+                    generic_kind,
+                    failed_due_to_placeholders: _,
```

Comment:
```suggestion
                    failed_due_to_placeholders: false,
```
or entirely split the type test error variant if they are always treated separately

Reply:
They are received separately (which is awkward), but the generating code is much simpler if this is just an aspect. I initially wrote them as separate, but having them in one variant simplified the code a lot in other places.